File: java/url-shorter/src/main/java/com/scdt/urlshorter/utils/SnowFlake.java
package com.scdt.urlshorter.utils;

import java.lang.management.ManagementFactory;
import java.lang.management.RuntimeMXBean;
import java.net.NetworkInterface;
import java.net.SocketException;
import java.util.Enumeration;

/**
 * 分布式雪花算法工具类
 *
 * @author mingo
 */
public class SnowFlake
{
    private final static long TWEPOCH = 12888349746579L;

    /**
     * 机器标识位数
     */
    private final static long WORKER_ID_BITS = 5L;

    /**
     * 数据中心标识位数
     */
    private final static long DATACENTER_ID_BITS = 5L;

    /**
     * 毫秒内自增位数
     */
    private final static long SEQUENCE_BITS = 12L;

    /**
     * 机器ID偏左移12位
     */
    private final static long WORKER_ID_SHIFT = SEQUENCE_BITS;

    /**
     * 数据中心ID左移17位
     */
    private final static long DATACENTER_ID_SHIFT = SEQUENCE_BITS + WORKER_ID_BITS;

    /**
     * 时间毫秒左移22位
     */
    private final static long TIMESTAMP_LEFT_SHIFT = SEQUENCE_BITS + WORKER_ID_BITS + DATACENTER_ID_BITS;

    /**
     * sequence掩码，确保sequence不会超出上限
     */
    private final static long SEQUENCE_MASK = ~(-1L << SEQUENCE_BITS);

    /**
     * 上次时间戳
     */
    private static long lastTimestamp = -1L;

    /**
     * 序列
     */
    private long sequence = 0L;

    /**
     * 服务器ID
     */
    private long workerId = 1L;

    private static final long WORKER_MASK = ~(-1L << WORKER_ID_BITS);

    /**
     * 进程编码
     */
    private long processId = 1L;

    private static final long PROCESS_MASK = ~(-1L << DATACENTER_ID_BITS);

    private static final SnowFlake SNOW_FLAKE;

    static
    {
        SNOW_FLAKE = new SnowFlake();
    }

    /**
     * 生成ID序列
     *
     * @return ID
     */
    public static synchronized long nextId()
    {
        return SNOW_FLAKE.getNextId();
    }

    private SnowFlake()
    {
        //获取机器编码
        this.workerId = this.getMachineNum();
        //获取进程编码
        RuntimeMXBean runtimeBean = ManagementFactory.getRuntimeMXBean();
        this.processId = Long.parseLong(runtimeBean.getName().split("@")[0]);

        //避免编码超出最大值
        this.workerId = workerId & WORKER_MASK;
        this.processId = processId & PROCESS_MASK;
    }

    public synchronized long getNextId()
    {
        //获取时间戳
        long timestamp = timeGenerator();
        //如果时间戳小于上次时间戳则报错
        if (timestamp < lastTimestamp)
        {
            try
            {
                throw new Exception("Clock moved backwards.  Refusing to generate id for "
                        + (lastTimestamp - timestamp)
                        + " milliseconds");
            }
            catch (Exception e)
            {
                e.printStackTrace();
            }
        }
        //如果时间戳与上次时间戳相同
        if (lastTimestamp == timestamp)
        {
            // 当前毫秒内，则+1，与sequenceMask确保sequence不会超出上限
            sequence = (sequence + 1) & SEQUENCE_MASK;
            if (sequence == 0)
            {
                // 当前毫秒内计数满了，则等待下一秒
                timestamp = tilNextMillis(lastTimestamp);
            }
        }
        else
        {
            sequence = 0;
        }
        lastTimestamp = timestamp;
        // ID偏移组合生成最终的ID，并返回ID
        return ((timestamp - TWEPOCH) << TIMESTAMP_LEFT_SHIFT)
                | (processId << DATACENTER_ID_SHIFT)
                | (workerId << WORKER_ID_SHIFT)
                | sequence;
    }

    /**
     * 再次获取时间戳直到获取的时间戳与现有的不同
     *
     * @param lastTimestamp 时间戳
     * @return 下一个时间戳
     */
    private long tilNextMillis(final long lastTimestamp)
    {
        long timestamp = this.timeGenerator();
        while (timestamp <= lastTimestamp)
        {
            timestamp = this.timeGenerator();
        }
        return timestamp;
    }

    private long timeGenerator()
    {
        return System.currentTimeMillis();
    }

    /**
     * 获取机器编码
     *
     * @return 机器编码
     */
    private long getMachineNum()
    {
        long machinePiece = 0;
        StringBuilder stringBuilder = new StringBuilder();
        Enumeration<NetworkInterface> interfaceEnumeration = null;
        try
        {
            interfaceEnumeration = NetworkInterface.getNetworkInterfaces();
        }
        catch (SocketException e1)
        {
            e1.printStackTrace();
        }

        if (interfaceEnumeration == null)
        {
            return machinePiece;
        }
        while (interfaceEnumeration.hasMoreElements())
        {
            NetworkInterface ni = interfaceEnumeration.nextElement();
            stringBuilder.append(ni.toString());
        }
        machinePiece = stringBuilder.toString().hashCode();
        return machinePiece;
    }
}

File: backend/src/utils/Snowflake.ts
import { convert10to62 } from "./RadixUtil";

class Snowflake {
  private twepoch: bigint
  private sequenceBits: bigint
  private timestampLeftShift: bigint
  private sequenceMask: bigint
  public lastTimestamp: bigint
  public sequence: bigint

  constructor() {

    /** 开始时间截 ：2023-03-08 16:49:23 */
    this.twepoch = 1678454599724n;

    /** 序列在id中占的位数 */
    this.sequenceBits = 8n;

    this.timestampLeftShift = this.sequenceBits;
    this.sequenceMask = -1n ^ (-1n << this.sequenceBits);

    /** 上次生成ID的时间截 */
    this.lastTimestamp = -1n;
    this.sequence = 0n;
  }

  public nextId() {
    var timestamp = this.timeGen();

    //如果当前时间小于上一次ID生成的时间戳，说明系统时钟回退过这个时候应当抛出异常
    if (timestamp < this.lastTimestamp) {
      throw new Error('Clock moved backwards. Refusing to generate id for ' +
        (this.lastTimestamp - timestamp));
    }

    //如果是同一时间生成的，则进行毫秒内序列
    if (this.lastTimestamp === timestamp) {
      this.sequence = (this.sequence + 1n) & this.sequenceMask;
      //毫秒内序列溢出
      if (this.sequence === 0n) {
        //阻塞到下一个毫秒,获得新的时间戳
        timestamp = this.tilNextMillis(this.lastTimestamp);
      }
    } else {
      //时间戳改变，毫秒内序列重置
      this.sequence = 0n;
    }

    //上次生成ID的时间截
    this.lastTimestamp = timestamp;

    //移位并通过或运算拼到一起组成64位的ID
    let result = ((timestamp - this.twepoch) << this.timestampLeftShift) | this.sequence
    return result;
  };

  /**
   * 阻塞到下一个毫秒，直到获得新的时间戳
   * @param lastTimestamp 上次生成ID的时间截
   * @return 当前时间戳
   */
  private tilNextMillis(lastTimestamp: bigint) {
    var timestamp = this.timeGen();
    while (timestamp <= lastTimestamp) {
      timestamp = this.timeGen();
    }
    return timestamp;
  };

  /**
   * 返回以毫秒为单位的当前时间
   * @return 当前时间(毫秒)
   */
  private timeGen() {
    return BigInt(Date.now());
  };

  public snowflakeGenerator() {
    return convert10to62(this.nextId())
  }
}

const snowflake = new Snowflake();

export default snowflake;